File: clip-watcher-menu.py
#!/usr/bin/env python3
"""
clip-watcher menu bar app
Provides a macOS menu bar interface for clip-watcher.sh
"""

import rumps
import subprocess
import os
import threading
from pathlib import Path

# Path to clip-watcher.sh
SCRIPT_PATH = Path.home() / "scripts" / "clip-watcher.sh"

# Menu bar appearance
APP_NAME = "Clip Watcher"
ICON_RUNNING = "▶"      # Play symbol when running
ICON_STOPPED = "⏹"      # Stop symbol when stopped  
ICON_UNKNOWN = "?"       # Unknown state
ICON_STARTING = "..."    # Starting up

class ClipWatcherApp(rumps.App):
    def __init__(self):
        super().__init__(APP_NAME, quit_button=None)
        self.script_path = str(SCRIPT_PATH)
        self.update_status()
        
        # Build menu
        self.menu = [
            rumps.MenuItem("Start", callback=self.start),
            rumps.MenuItem("Stop", callback=self.stop),
            None,  # Separator
            rumps.MenuItem("Status", callback=self.status),
            rumps.MenuItem("Last Link", callback=self.last_link),
            None,  # Separator
            rumps.MenuItem("Tail Log", callback=self.tail_log),
            rumps.MenuItem("Clear Processed", callback=self.clear_processed),
            None,  # Separator
            rumps.MenuItem("Quit", callback=self.quit),
        ]
        
        # Start status update timer
        self.timer = rumps.Timer(self.update_status, 5)
        self.timer.start()
    
    def run_command(self, args, callback=None):
        """Run a clip-watcher command in background"""
        def _run():
            try:
                result = subprocess.run(
                    [self.script_path] + args,
                    capture_output=True,
                    text=True,
                    timeout=30
                )
                if callback:
                    callback(result.stdout, result.stderr, result.returncode)
            except subprocess.TimeoutExpired:
                if callback:
                    callback("", "Command timed out", 1)
            except Exception as e:
                if callback:
                    callback("", str(e), 1)
        
        threading.Thread(target=_run, daemon=True).start()
    
    def update_status(self, timer=None):
        """Update menu bar icon based on watcher status"""
        try:
            pid_file = Path.home() / "Movies" / "Captures" / "encoded" / ".clip-watcher.pid"
            if pid_file.exists():
                pid = pid_file.read_text().strip()
                # Check if process is running
                result = subprocess.run(
                    ["kill", "-0", pid],
                    capture_output=True,
                    timeout=5
                )
                if result.returncode == 0:
                    self.title = ICON_RUNNING
                    return
            self.title = ICON_STOPPED
        except Exception:
            self.title = ICON_UNKNOWN
    
    @rumps.clicked("Start")
    def start(self, _):
        def on_complete(stdout, stderr, returncode):
            if returncode == 0:
                rumps.notification("Clip Watcher", "Started", "Watcher is now running")
            else:
                rumps.notification("Clip Watcher", "Start Failed", stderr or "Check logs")
            self.update_status()
        
        self.run_command([], on_complete)
    
    @rumps.clicked("Stop")
    def stop(self, _):
        def on_complete(stdout, stderr, returncode):
            rumps.notification("Clip Watcher", "Stopped", stdout.strip() if stdout else "Watcher stopped")
            self.update_status()
        
        self.run_command(["stop"], on_complete)
    
    @rumps.clicked("Status")
    def status(self, _):
        def on_complete(stdout, stderr, returncode):
            if stdout:
                # Show status in a dialog
                rumps.alert(
                    title="Clip Watcher Status",
                    message=stdout.strip(),
                    ok="OK"
                )
            else:
                rumps.notification("Clip Watcher", "Status", stderr or "Could not get status")
        
        self.run_command(["status"], on_complete)
    
    @rumps.clicked("Last Link")
    def last_link(self, _):
        def on_complete(stdout, stderr, returncode):
            if returncode == 0 and stdout:
                # Copy to clipboard
                import subprocess
                subprocess.run(["pbcopy"], input=stdout.strip().encode(), check=True)
                rumps.notification("Clip Watcher", "Link Copied", stdout.strip())
            else:
                rumps.notification("Clip Watcher", "No Links", stderr or "No links recorded yet")
        
        self.run_command(["last"], on_complete)
    
    @rumps.clicked("Tail Log")
    def tail_log(self, _):
        log_file = Path.home() / "Movies" / "Captures" / "encoded" / "clip-watcher.log"
        if log_file.exists():
            # Open log file in default text editor
            subprocess.run(["open", "-a", "TextEdit", str(log_file)])
        else:
            rumps.notification("Clip Watcher", "No Log", "Log file not found")
    
    @rumps.clicked("Clear Processed")
    def clear_processed(self, _):
        response = rumps.alert(
            title="Clear Processed Log",
            message="This will allow all clips to be re-queued. Continue?",
            ok="Clear",
            cancel="Cancel"
        )
        if response == 1:  # OK clicked
            def on_complete(stdout, stderr, returncode):
                rumps.notification("Clip Watcher", "Cleared", "Processed log cleared")
            
            self.run_command(["clear"], on_complete)
    
    @rumps.clicked("Quit")
    def quit(self, _):
        rumps.quit_application()


if __name__ == "__main__":
    ClipWatcherApp().run()

File: ClipWatcherMenu.swift
#!/usr/bin/swift

import Cocoa

// ── Configuration ───────────────────────────────────────────────────────────

let CLIP_PATH = NSString("~/Documents/GitHub/ncembed/clip").expandingTildeInPath
let CONFIG_DIR = NSString("~/.config/clip-watcher").expandingTildeInPath
let CONFIG_FILE = "\(CONFIG_DIR)/config.json"
let TEMP_DIR = NSString("~/Movies/Captures/encoded").expandingTildeInPath
let PID_FILE = "\(TEMP_DIR)/.clip-watcher.pid"
let LOG_FILE = "\(TEMP_DIR)/clip-watcher.log"

struct Config: Codable {
    var watchDir: String
    var nextcloudURL: String
    var nextcloudUser: String
    var nextcloudPass: String
    var uploadPath: String
    var ncembedDomain: String
    var useNcembed: Bool
    var sambaShares: [String]
}

func loadConfig() -> Config? {
    guard FileManager.default.fileExists(atPath: CONFIG_FILE) else { return nil }
    guard let data = try? Data(contentsOf: URL(fileURLWithPath: CONFIG_FILE)) else { return nil }
    return try? JSONDecoder().decode(Config.self, from: data)
}

// ── Menu Bar App ────────────────────────────────────────────────────────────

class AppDelegate: NSObject, NSApplicationDelegate {
    var statusItem: NSStatusItem!
    var timer: Timer?
    var statusMenuItem: NSMenuItem!
    
    func applicationDidFinishLaunching(_ notification: Notification) {
        NSApp.setActivationPolicy(.accessory)
        
        setupStatusBar()
        updateStatus()
        
        timer = Timer.scheduledTimer(withTimeInterval: 3.0, repeats: true) { _ in
            self.updateStatus()
        }
    }
    
    func setupStatusBar() {
        statusItem = NSStatusBar.system.statusItem(withLength: NSStatusItem.variableLength)
        
        if let button = statusItem.button {
            button.title = "⏹ Clip"
        }
        
        let menu = NSMenu()
        
        statusMenuItem = NSMenuItem(title: "Status: Checking...", action: nil, keyEquivalent: "")
        statusMenuItem.isEnabled = false
        menu.addItem(statusMenuItem)
        
        menu.addItem(NSMenuItem.separator())
        menu.addItem(NSMenuItem(title: "Start", action: #selector(startWatcher), keyEquivalent: "s"))
        menu.addItem(NSMenuItem(title: "Stop", action: #selector(stopWatcher), keyEquivalent: "x"))
        menu.addItem(NSMenuItem.separator())
        menu.addItem(NSMenuItem(title: "Copy Last Link", action: #selector(copyLastLink), keyEquivalent: "c"))
        menu.addItem(NSMenuItem(title: "Tail Log", action: #selector(tailLog), keyEquivalent: "l"))
        menu.addItem(NSMenuItem.separator())
        menu.addItem(NSMenuItem(title: "Quit", action: #selector(quitApp), keyEquivalent: "q"))
        
        statusItem.menu = menu
    }
    
    func updateStatus() {
        let running = isWatcherRunning()
        
        if let button = statusItem.button {
            button.title = running ? "▶ Clip" : "⏹ Clip"
        }
        statusMenuItem.title = running ? "Status: Running" : "Status: Stopped"
    }
    
    func isWatcherRunning() -> Bool {
        guard let pidStr = try? String(contentsOfFile: PID_FILE, encoding: .utf8),
              let pid = Int32(pidStr.trimmingCharacters(in: .whitespacesAndNewlines)) else {
            return false
        }
        return kill(pid, 0) == 0
    }
    
    // ── Actions ─────────────────────────────────────────────────────────────
    
    @objc func startWatcher() {
        runClip("start")
        DispatchQueue.main.asyncAfter(deadline: .now() + 2) { self.updateStatus() }
    }
    
    @objc func stopWatcher() {
        runClip("stop")
        DispatchQueue.main.asyncAfter(deadline: .now() + 1) { self.updateStatus() }
    }
    
    @objc func copyLastLink() {
        let output = runClipWithOutput("last")
        if output.contains("Copied:") {
            let parts = output.components(separatedBy: "Copied: ")
            if parts.count > 1 {
                let url = parts[1].components(separatedBy: "  ").first ?? ""
                NSPasteboard.general.clearContents()
                NSPasteboard.general.setString(url, forType: .string)
                notify("Link Copied", url)
            }
        } else {
            notify("No Links", output.isEmpty ? "No uploads recorded yet" : output)
        }
    }
    
    @objc func tailLog() {
        if FileManager.default.fileExists(atPath: LOG_FILE) {
            NSWorkspace.shared.open(URL(fileURLWithPath: LOG_FILE))
        } else {
            notify("No Log", "Log file not found")
        }
    }
    
    @objc func quitApp() {
        NSApplication.shared.terminate(nil)
    }
    
    // ── Helpers ─────────────────────────────────────────────────────────────
    
    @discardableResult
    func runClip(_ args: String...) -> Int32 {
        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: "/bin/zsh")
        proc.arguments = ["-l", "-c", "\(CLIP_PATH) \(args.joined(separator: " "))"]
        try? proc.run()
        proc.waitUntilExit()
        return proc.terminationStatus
    }
    
    func runClipWithOutput(_ args: String...) -> String {
        let proc = Process()
        let pipe = Pipe()
        proc.executableURL = URL(fileURLWithPath: "/bin/zsh")
        proc.arguments = ["-l", "-c", "\(CLIP_PATH) \(args.joined(separator: " "))"]
        proc.standardOutput = pipe
        proc.standardError = pipe
        try? proc.run()
        proc.waitUntilExit()
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        return String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    }
    
    func notify(_ title: String, _ message: String) {
        let t = title.replacingOccurrences(of: "\"", with: "\\\"")
        let m = message.replacingOccurrences(of: "\"", with: "\\\"")
        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: "/usr/bin/osascript")
        proc.arguments = ["-e", "display notification \"\(m)\" with title \"\(t)\""]
        try? proc.run()
    }
}

// ── Main ────────────────────────────────────────────────────────────────────

let app = NSApplication.shared
let delegate = AppDelegate()
app.delegate = delegate
app.run()
